src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryGuidConverter.cs

```diff
@@ -0,0 +1,186 @@
+                {
+                    return false;
+                }
+
+                if (typeToConvert.GetGenericTypeDefinition() != typeof(Dictionary<,>))
+                {
+                    return false;
+                }
+
+                return (typeToConvert.GetGenericArguments()[0] == typeof(Guid));
+            }
+
+            public override JsonConverter CreateConverter(Type type, JsonSerializerOptions options)
+            {
+                Type valueType = type.GetGenericArguments()[1];
+
+                JsonConverter converter = (JsonConverter)Activator.CreateInstance(
+                    typeof(DictionaryGuidConverterInner<>).MakeGenericType(new Type[] { valueType }),
+                    BindingFlags.Instance | BindingFlags.Public,
+                    binder: null,
+                    args: new object[] { options },
+                    culture: null);
+
+                return converter;
+            }
+
+            private class DictionaryGuidConverterInner<TValue> : JsonConverter<Dictionary<Guid, TValue>>
+            {
+                private readonly JsonConverter<TValue> _valueConverter;
+
+                public DictionaryGuidConverterInner(JsonSerializerOptions options)
+                {
+                    // For performance, use the existing converter if available.
+                    _valueConverter = (JsonConverter<TValue>)options.GetConverter(typeof(TValue));
+                }
+
+                public override Dictionary<Guid, TValue> Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+                {
+                    if (reader.TokenType != JsonTokenType.StartObject)
+                    {
+                        throw new JsonException();
+                    }
+
+                    Dictionary<Guid, TValue> value = new Dictionary<Guid, TValue>();
+
+                    while (reader.Read())
+                    {
+                        if (reader.TokenType == JsonTokenType.EndObject)
+                        {
+                            return value;
+                        }
+
+                        // Get the key.
+                        if (reader.TokenType != JsonTokenType.PropertyName)
+                        {
+                            throw new JsonException();
+                        }
+
+                        string propertyName = reader.GetString();
+
+                        // Parse guid in "D" format: 00000000-0000-0000-0000-00000000000
+                        if (!Guid.TryParseExact(propertyName, format:"D", out Guid result))
+                        {
+                            throw new JsonException($"Unable to convert \"{propertyName}\" to a Guid.");
+                        }
+
+                        // Get the value.
+                        TValue v;
+                        if (_valueConverter != null)
+                        {
+                            reader.Read();
```

Comment:
Interesting that the reader needs to be set to past the value before passing it into the `_valueConverter`. Looking at this from outside in, I would have expected the converter to do a single read itself.